library/core/src/lib.rs

```diff
@@ -20,11 +20,14 @@
 // FIXME: Fill me in with more detail when the interface settles
 //! This library is built on the assumption of a few existing symbols:
 //!
-//! * `memcpy`, `memcmp`, `memset`, `strlen` - These are core memory routines which are
-//!   often generated by LLVM. Additionally, this library can make explicit
-//!   calls to these functions. Their signatures are the same as found in C.
-//!   These functions are often provided by the system libc, but can also be
-//!   provided by the [compiler-builtins crate](https://crates.io/crates/compiler_builtins).
+//! * `memcpy`, `memmove`, `memcmp`, `memset`, `strlen` - These are core memory routines which are
+//!   generated by Rust codegen backends. Additionally, this library can make explicit calls to
+//!   `strlen`. Their signatures are the same as found in C, but there are extra assumptions about
+//!   their semantics: For `memcpy`, `memmove`, `memcmp`, and `memset`, if the `n` parameter is 0,
+//!   the function is assumed to not be UB. Furthermore, for `memcpy`, if source and target pointer
+//!   are equal, the function is assumed to not be UB.
```

Comment:
I think it's worth pointing out that these assumptions come from LLVM and are shared by GCC. Otherwise users could be surprised that Rust has such "strong requirements", but putting them into perspective with LLVM and GCC shows that it's really not that strong.

Reply:
The memcmp assumption is only ours, AFAIK.

Does GCC document their assumptions somewhere? For LLVM, should I link to that phabricator review?

Reply:
ideally it would link to the LLVM docs after those are merged

Reply:
Well the LLVM PR has been sitting for almost 3 years so I wouldn't want to wait for that.^^